Improvements to horizontal spacing

mike-spa opened the pull request:
This PR contains several improvements that @oktophonie and I have been working on. Apart from posssible bug fixes or small corrections, this may be somewhat the "definitive" horizontal spacing for MU4.

The PR is split into 3 commits which are logically separate and may be also reviewed and merged separately, if needed.

1. I've rolled back the `acceptanceRange` concept. This used to let systems be initially slightly wider than the margins, and justify them by squeezing rather than stretching. The idea is good, but it is still too risky and it can break under some edge cases. Hopefully I can bring it back in future.

2. I've developed better logic to prevent the spacing of long notes to explode in the presence of very short ones. The previous logic was hacky and only worked with one specific spacing formula. The new one is mathematically solid, and works in general regardless of which spacing formula one may choose.

3. An overall improvement of the spacing algorithm.
- The spacing formula has been changed in favor of a better one.
- Some parameters, namely `minNoteDistance`, `stretch`, and `measureSpacing` interact differently with the spacing algorithm. The most significant change is `measureSpacing`. This parameter now does not act as an overall multiplier, but affects the *ratio* between the space assigned to longer vs shorter notes. This also required a minor UI change. The "Measure -> Spacing" style setting has been renamed to "Measure -> Spacing curve". Now this parameter can only take values between 1 and 2, where 1 corresponds to a flat spacing curve (meaning all notes receive equal space regardless of duration), and 2 corresponds to strict time-space layout. The default value of "Measure -> Spacing curve" is now 1.5. This required to change all the template files so that they load with the correct value. Additionally, I'm making sure that the value is reset to default for any score migrating from an older version of Musescore.

MarcSabatella commented:
The "spacing curve" change is exciting, strict proportional spacing is a semi-common request!

asattely commented on src/engraving/libmscore/measure.cpp:
```diff
@@ -4043,12 +4037,22 @@ float Measure::durationStretch(Fraction curTicks, const Fraction minTicks) const
+    double ratio = double(curTicks.ticks()) / minSysTicks;
+    if (maxSysRatio > maxRatio) {
+        double A = minSysTicks * (maxRatio - 1) / (maxSysTicks - minSysTicks);
+        double B = (maxSysTicks - maxRatio * minSysTicks) / (maxSysTicks - minSysTicks);
```
for these three lines, it's better to use parentheses to make sure the order of operations is what you want it to be. There have been a number of bugs I've been fixing recently that stem from people being unclear about the order of operations, so it's better to write
`((maxSysTicks - maxRatio) * minSysTicks) / (maxSysTicks - minSysTicks);`, for example

asattely commented on src/engraving/libmscore/measure.cpp:
```diff
@@ -4057,11 +4061,13 @@ float Measure::durationStretch(Fraction curTicks, const Fraction minTicks) const
+    if (minSysTicks > longNoteThreshold) {
         // Avoids long notes being too narrow in the absense of shorter notes.
-        str = str * (1 - 0.5 + 0.5 * sqrt(qreal(minTicks.ticks()) / qreal(longNoteThreshold.ticks())));
+        str = str * (1 - 0.6 + 0.6 * sqrt(minSysTicks / longNoteThreshold));
```
a few magic numbers in here, it might help to make named constants to help people who want to work on this code in the future! Or at least a comment explaining what these `0.6` values are.

Jojo-Schmitz commented:
The mtests failures need to get fixed, most probably by adjusting the ref files

asattely commented:
> I guess I should probably just update all the vtest files to have the parameter already set to 1.5?
As far as I know, you don't need to update vtests manually! Someone with the ability will add the 'vtests' label to it as long as all the vtests show expected results

Jojo-Schmitz commented:
@shoogle ?

mike-spa commented:
> edit: Oh, I see what you mean. Yeah potentially you will have to fix the files the vtests use. Definitely ask someone like Peter about that

Yeah sorry I expressed myself quite poorly. What's confusing me is that if I manually open any (old) vtest file with MU4, the migration routine will update the "Measure -> Spacing" style parameter to its new default value, as it should. But this doesn't seem to happen during the actual running of the vtests. They seem to run without updating the parameter. So if we now put the vtest label, the new reference images would be generated with the wrong style value.

mike-spa commented:
I've added a further commit (again, it can be reviewed and merged separately if needed). Besides the 3 previous points, we now have:

4. Improvements of the spacing calculations when rythms overlap non-trivially. The best example is polyrythms. We go from this:
![Screenshot from 2022-03-03 18-44-01](https://user-images.githubusercontent.com/93707756/156731086-f1f1b9f7-fbef-4eb2-85a7-5ce8a33af04e.png)
to this:
![Screenshot from 2022-03-03 18-37-52](https://user-images.githubusercontent.com/93707756/156731133-9fb61e6a-6501-4397-a62c-6afea3d06f21.png)
This ensures correct spacing as long as there is enough horizontal space. The proportions will still be distorted when notes start coming close to each other. This, however, will be solved when we let notes from different voices overlap vertically. This will require a complete overhaul of the collision checking algorithm.

Besides that, I've also implemented @asattely 's suggestions. Next, I'll look into those vtests and also fix the utests failures.

mike-spa commented:
The vtests now are all loading the correct default values of `measureSpacing = 1.5` (thanks @cbjeukendrup and @asattely for the great help). After some minor bug fixes, they now correctly show the updated horizontal spacing in all cases so, pending @oktophonie 's review, I think the vtest label is in order. utests (and there were *a lot* of them) have been also fixed. 
I'm now not planning to make any additional change, so it is in principle ready for review.

vpereverzev commented on src/engraving/libmscore/segment.cpp:
```diff
@@ -2691,4 +2691,28 @@ qreal Segment::minHorizontalDistance(Segment* ns, bool systemHeaderGap) const
+    for (auto elem : elist()) {
+        if (!elem || !elem->staff()->show() || !elem->isChordRest()) {
+            continue;
+        } else {
```
no need for "else" if you have "continue" above

vpereverzev commented on src/engraving/layout/layoutsystem.cpp:
```diff
@@ -171,9 +171,10 @@ System* LayoutSystem::collectSystem(const LayoutOptions& options, LayoutContext&
         // collect at least one measure and the break
 
-        double acceptanceRange = system->isSqueezable() ? 1.025 : 1; // A value slightly larger than 1 allows systems
+        //double acceptanceRange = system->isSqueezable() ? 1.025 : 1; // A value slightly larger than 1 allows systems
```
don't leave it like this, better to remove this line if you don't need it

mike-spa commented:
> Has this been replaced by #10934?

Correct. I'm gonna close this one.

mike-spa commented:
Replaced by [#10934](https://github.com/musescore/MuseScore/pull/10934)
